tools/checksum_bitflip/checksum_bitflip.c

```diff
@@ -0,0 +1,107 @@
+/*-
+ * Public Domain 2024-present MongoDB, Inc.
+ *
+ * This is free and unencumbered software released into the public domain.
+ *
+ * Anyone is free to copy, modify, publish, use, compile, sell, or
+ * distribute this software, either in source code form or as a compiled
+ * binary, for any purpose, commercial or non-commercial, and by any
+ * means.
+ *
+ * In jurisdictions that recognize copyright laws, the author or authors
+ * of this software dedicate any and all copyright interest in the
+ * software to the public domain. We make this dedication for the benefit
+ * of the public at large and to the detriment of our heirs and
+ * successors. We intend this dedication to be an overt act of
+ * relinquishment in perpetuity of all present and future rights to this
+ * software under copyright law.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+ * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#include <test_util.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+
+/*
+ * This program takes a file and a checksum and checks whether the checksum
+ * matches the contents of the file, or if any checksum produced by flipping
+ * a single bit of the file contents matches.
+ *
+ * This intended for use when debugging WT checksum mismatches when we suspect
+ * that the issue is the result of faulty hardware causing bit flips in memory.
+ *
+ * Usage:
+ *     checksum_bitflip cksum filename
+ *
+ * The checksum argument should be a hex string.
+ */
+
+int
+main(int argc, char *argv[])
+{
+    char *buffer, *filename;
+    uint8_t mask;
+    int bit, fd;
+    size_t byte, size;
+    ssize_t io_bytes;
+    uint32_t cksum_target;
+    int64_t tmp;
+    struct stat statbuf;
+
+    if (argc < 3) {
+        printf ("Usage: %s checksum filename\n", argv[0]);
+        exit(1);
+    }
+
+    tmp = strtol(argv[1], NULL, 16);
+    if ((tmp ^ (tmp & 0xFFFFFFFF)) != 0) {
+        fprintf (stderr, "Target checksum must be 32-bits\n");
+        exit(1);
+    }
+    cksum_target = (uint32_t) tmp;
+
+    filename = argv[2];
+    fd = open(filename, O_RDONLY);
+    fstat(fd, &statbuf);
+    size = (size_t) statbuf.st_size;
+    buffer = dmalloc(size);
+
+    io_bytes = read(fd, buffer, size);   
+    if ((size_t) io_bytes != size) {
+        printf ("Read of %s returned %zd\n", filename, io_bytes);
+        exit(1);
+    }
```

Comment:
Hmm, I thought WT checksummed each block/page it writes, not the entire file. So I thought this would be checksumming 32K chunks or whatever size.

Reply:
Ah. I wasn't clear about this. I don't point this at a `*.wt` file. Instead I take the hex dump we generate on checksum mismatch, and convert that into binary. Then run this tool on the resulting file. The advantage is that you can do that on customer failures where we don't have access to the WiredTiger file.

I've added text to the opening comment explaining this.

But I like the idea of being able to specify an offset and size as optional parameters so you could also use this on a WiredTiger file. I'll create a follow-up ticket for that.

Reply:
Meta-comment. Some of the issues here are because we apparently don't run clang-format on things in the `tools/` directory. It looks like this is the first C program we've added there.